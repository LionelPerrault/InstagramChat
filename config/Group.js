// HANDY METHODS FOR GROUP ROUTES

const db = require('./db'),
  { rmdir } = require('fs'),
  { promisify } = require('util'),
  root = process.cwd(),
  { deletePost } = require('./Post'),
  { DeleteAllOfFolder } = require('handy-image-processor'),
  { intersectionBy } = require('lodash')

/**
 * Returns what of a group
 * @param {String} what Get what eg. name
 * @param {Number} group Group ID
 */
const getWhatOfGrp = async (what, group) => {
  let s = await db.query(`SELECT ${what} FROM groups WHERE group_id=?`, [group])
  return s.length == 0 ? '' : s[0][what]
}

/**
 * Deletes group
 * @param {Number} group GroupID
 */
const deleteGroup = async group => {
  let posts = await db.query('SELECT post_id FROM posts WHERE group_id=?', [
      group,
    ]),
    dltDir = promisify(rmdir)

  for (let p of posts) {
    await deletePost({ post: p.post_id, when: 'group' })
  }

  await db.query('DELETE FROM notifications WHERE group_id=?', [group])
  await db.query('DELETE FROM group_members WHERE group_id=?', [group])
  await db.query('DELETE FROM groups WHERE group_id=?', [group])

  DeleteAllOfFolder(`${root}/dist/groups/${group}/`)
  await dltDir(`${root}/dist/groups/${group}`)
}

/**
 * Returns whether user joined group
 * @param {Number} user UserID
 * @param {Number} group GroupID
 */
const joinedGroup = async (user, group) => {
  let is = await db.query(
    'SELECT COUNT(grp_member_id) AS joined FROM group_members WHERE member=? AND group_id=? LIMIT 1',
    [user, group]
  )
  return db.tf(is[0].joined)
}

/**
 * Returns mutual users of group members and user
 * @param {Number} user UserID
 * @param {Number} group GroupID
 */
const mutualGroupMembers = async (user, group) => {
  let myFollowings = await db.query(
      'SELECT follow_system.follow_to AS user, follow_system.follow_to_username AS username FROM follow_system WHERE follow_system.follow_by=?',
      [user]
    ),
    grpMembers = await db.query(
      'SELECT group_members.member AS user, users.username AS username FROM group_members, users WHERE group_id = ? AND group_members.member = users.id ORDER BY group_members.joined_group DESC',
      [group]
    ),
    mutuals = intersectionBy(myFollowings, grpMembers, 'user')

  return mutuals
}

module.exports = {
  getWhatOfGrp,
  deleteGroup,
  joinedGroup,
  mutualGroupMembers,
}
