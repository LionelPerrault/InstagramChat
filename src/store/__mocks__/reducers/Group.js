export default {
  group_details: {
    group_id: 11,
    name: 'a groupss',
    bio: '#random group',
    admin: 24,
    admin_username: 'takkar',
    created: '1480114098767',
    group_type: 'public',
  },
  userGroups: [
    {
      admin: 7,
      group_id: 11,
      joined: true,
      joined_group: '1480114098767',
      member: 24,
      name: 'a groupss',
    },
  ],
  mutualMembers: [
    {
      user: 7,
      username: 'ghalib',
    },
  ],
  newestMembers: [
    {
      user: 7,
      username: 'ghalib',
    },
  ],
  usersToInvite: [
    {
      follow_id: 302,
      follow_to: 16,
      username: 'zayn',
      firstname: 'Zayn',
      surname: 'Malik',
    },
  ],
  usersToMakeAdmin: [
    {
      grp_member_id: 45,
      member: 34,
      username: 'ragnar',
      firstname: 'Ragnar',
      surname: 'Lothbrok',
    },
  ],
  members: [
    {
      added_by: 24,
      added_by_username: 'takkar',
      firstname: 'Mirza',
      group_id: 11,
      grp_member_id: 58,
      joined_group: '1480114098767',
      member: 7,
      mutualUsersCount: 1,
      surname: 'Ghalib',
      username: 'ghalib',
    },
  ],
  joined: false,
}
