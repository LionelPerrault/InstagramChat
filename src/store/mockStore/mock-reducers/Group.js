export default {
  group_details: {
    group_id: 11,
    name: 'a groupss',
    bio: '#random group',
    admin: 24,
    admin_username: 'takkar',
    created: '1518016737587',
    group_type: 'private',
  }
}
