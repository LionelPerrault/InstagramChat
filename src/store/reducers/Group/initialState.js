export default {
  group_details: {
    group_id: null,
    name: '',
    bio: '',
    admin: null,
    admin_username: '',
    created: '',
    group_type: '',
    postsCount: 0,
  },
  joined: false,
  members: [],
  mutualMembers: [],
  newestMembers: [],
  userGroups: [],
  usersToInvite: [],
  usersToMakeAdmin: [],
}
