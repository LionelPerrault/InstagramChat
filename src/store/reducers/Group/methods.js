export const update = (group, { name, bio, group_type }) => {
  let updated = {
    ...group,
    name,
    bio,
    group_type,
  }
  return updated
}

export const remMember = (members, member_id) =>
  members.filter(m => m.grp_member_id != member_id)

export const leftGroup = (groups, group_id) =>
  groups.filter(g => g.group_id != group_id)
