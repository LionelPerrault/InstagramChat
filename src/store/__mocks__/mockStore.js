import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'

// mock-reducers
import User from './reducers/User'
import Notification from './reducers/Notification'
import Group from './reducers/Group'
import Follow from './reducers/Follow'
import Explore from './reducers/Explore'
import Post from './reducers/Post'
import Hashtag from './reducers/Hashtag'
import Setting from './reducers/Setting'
import Message from './reducers/Message'

const initialState = {
  User,
  Notification,
  Group,
  Follow,
  Explore,
  Post,
  Hashtag,
  Setting,
  Message,
}

const middlewares = [thunk]
const createStore = configureStore(middlewares)
const mockStore = createStore(initialState)

export default mockStore
