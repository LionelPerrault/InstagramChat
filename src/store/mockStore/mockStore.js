import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import logger from 'redux-logger'
import promise from 'redux-promise-middleware'

// mock-reducers
import User from './mock-reducers/User'
import Notification from './mock-reducers/Notification'
import Group from './mock-reducers/Group'
import Follow from './mock-reducers/Follow'
import Explore from './mock-reducers/Explore'

const initialState = {
  User,
  Notification,
  Group,
  Follow,
  Explore
}

const middlewares = [
  promise(), thunk, logger
]
const createStore = configureStore(middlewares)
const mockStore = createStore(initialState)

export default mockStore
