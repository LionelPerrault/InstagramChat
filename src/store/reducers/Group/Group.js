/* eslint indent:0 */
/* eslint no-unreachable:0 */

import initialState from './initialState'
import * as methods from './methods'

export default (state = initialState, action) => {
  let py = action.payload

  switch (action.type) {
    case 'GET_GROUP_DETAILS':
      return { ...state, group_details: py }
      break

    case 'JOINED_GROUP':
      return { ...state, joined: py }
      break

    case 'TOGGLE_JOIN_GROUP':
      return { ...state, joined: py }
      break

    case 'UPDATE_GROUP':
      return {
        ...state,
        group_details: methods.update(state.group_details, py),
      }
      break

    case 'GET_GROUP_MEMBERS':
      return { ...state, members: py }
      break

    case 'REMOVE_MEMBER':
      return {
        ...state,
        members: methods.remMember(state.members, py),
      }
      break

    case 'GET_MUTUAL_AND_NEWEST_MEMBERS':
      return {
        ...state,
        mutualMembers: py.mutualMembers,
        newestMembers: py.newestMembers,
      }
      break

    case 'GET_USER_GROUPS':
      return { ...state, userGroups: py }
      break

    case 'LEFT_GROUP':
      return {
        ...state,
        userGroups: methods.leftGroup(state.userGroups, py),
      }
      break

    case 'GET_USERS_TO_INVITE':
      return { ...state, usersToInvite: py }
      break

    case 'GET_USERS_TO_MAKE_ADMIN':
      return { ...state, usersToMakeAdmin: py }
      break
  }

  return state
}
