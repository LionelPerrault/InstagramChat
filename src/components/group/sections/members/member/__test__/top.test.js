import React from 'react'
import MockDataElement from '../../../../../../utils/__mocks__/mock-dataElement'
import { Provider } from 'react-redux'
import mockStore from '../../../../../../store/__mocks__/mockStore'
import MembersTop from '../top'
import { create } from 'react-test-renderer'
import Group from '../../../../../../store/__mocks__/reducers/Group'

describe('MemberTop Component', () => {
  MockDataElement()

  const memberDetails = {
    member: 7,
    username: 'ghalib',
    firstname: 'Mirza',
    surname: 'Ghalib',
    mutualUsersCount: 0,
  }

  const comp = (extraProps = {}) => (
    <Provider store={mockStore}>
      <MembersTop
        memberDetails={{
          ...memberDetails,
          ...extraProps,
        }}
      />
    </Provider>
  )

  it('should match snapshot', () => {
    const tree = create(comp()).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should match snapshot and show mutual followers', () => {
    const tree = create(
      comp({
        mutualUsersCount: 2,
      })
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should match snapshot and show admin label when member is admin of the group', () => {
    Group.group_details.admin = 7
    const tree = create(comp()).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
