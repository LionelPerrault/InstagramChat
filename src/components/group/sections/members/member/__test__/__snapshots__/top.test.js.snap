// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemberTop Component should match snapshot 1`] = `
<div
  className="m_top"
>
  <img
    src="/users/7/avatar.jpg"
  />
  <div
    className="m_top_right"
  >
    <Link
      to="/profile/ghalib"
    >
      ghalib
    </Link>
    <span>
      Mirza Ghalib
    </span>
  </div>
</div>
`;

exports[`MemberTop Component should match snapshot and show admin label when member is admin of the group 1`] = `
<div
  className="m_top"
>
  <img
    src="/users/7/avatar.jpg"
  />
  <div
    className="m_top_right"
  >
    <Link
      to="/profile/ghalib"
    >
      ghalib
    </Link>
    <span
      className="grp_admin"
    >
      admin
    </span>
    <span>
      Mirza Ghalib
    </span>
  </div>
</div>
`;

exports[`MemberTop Component should match snapshot and show mutual followers 1`] = `
<div
  className="m_top"
>
  <img
    src="/users/7/avatar.jpg"
  />
  <div
    className="m_top_right"
  >
    <Link
      to="/profile/ghalib"
    >
      ghalib
    </Link>
    <span>
      2 mutual followers
    </span>
  </div>
</div>
`;
